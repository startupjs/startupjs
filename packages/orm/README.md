# startupjs racer-orm
ORM system for Racer.js and ShareDB

## What it does

Lets you automatically override your scope models
(created with `.at()` and `.scope()`) with the additional methods.

## Usage

Add the plugin:

```js
import Racer from 'racer'
import racerOrm from 'racer-orm'
Racer.use(racerOrm)
```

Then start adding the ORM entities to your model.
Each ORM Entity must be inherited from `Model.ChildModel`.

```js

import { Model } from 'racer'

class PlayerModel extends Model.ChildModel {
  alert (message) {
    this.set('alert', this.get('name') + ', ' + message)
    this.setDiff('showAlert', true)
  }
}

class GamesModel extends Model.ChildModel {
  async addNew (userId = 'system', params = {}) {
    let gameId = this.id()
    await this.addAsync('games', {
      name: 'Dummy Game',
      ...params,
      id: gameId,
      userId,
      playerIds: [],
      createdAt: Date.now()
    })
    return gameId
  }
}

class GameModel extends Model.ChildModel {
  async alertPlayers (message) {
    let playerIds = this.get('playerIds')
    let playersQuery = this.root.query('players', { _id: { $in: playerIds } })
    await this.subscribeAsync(playersQuery)
    for (let playerId of playersQuery.getIds()) {
      this.scope('players.' + playerId).alert(message)
    }
  }

  async addPlayer (userId, params = {}) {
    if (!userId) throw new Error('userId required')
    var playerId = this.id()
    await this.root.addAsync('players', {
      ...params,
      id: playerId,
      userId,
      createdAt: Date.now()
    })
    await this.pushAsync('playerIds', playerId)
    return playerId
  }
}

racer.orm('games', GamesModel)
racer.orm('games.*', GameModel)
racer.orm('players.*', PlayerModel)

// ...

async function main ($root) {
  let $games = $root.scope('games')
  let gameId = await $games.addNew('userId1', { name: 'Cool game' })
  let $game = $games.at(gameId)
  for (let userIds of ['userId1', 'userId2', 'userId3']) {
    await $game.addPlayer(userIds)
  }
  $game.alertPlayers('please join the game!')
}

// ...
```

## Factory

Sometimes you want to dynamically decide which ORM to use
based on the document's data. Factory let you do that.

Example:

```js
class BasePlayerModel extends Model.ChildModel {
  getColor () {
    throw new Error('Player color is unknown')
  }
}

class AlliedPlayerModel extends BasePlayerModel {
  getColor () {
    return 'blue'
  }
}

class RivalPlayerModel extends BasePlayerModel {
  getColor () {
    return 'red'
  }
}

function PlayerFactory ($player, $parent) {
  // $player here is going to be just a pure scoped model
  let playerTeamId = $player.get('teamId')
  let $root = $player.root
  let myTeamId = $root.get('_session.myTeamId')

  // you have to always pass `$parent` when manually
  // instantiating the ORM Entity
  if (!playerTeamId || !myTeamId) return new BasePlayerModel($parent)

  if (playerTeamId === myTeamId) {
    return new AlliedPlayerModel($parent)
  } else {
    return new RivalPlayerModel($parent)
  }
}
PlayerFactory.factory = true

racer.orm('players.*', PlayerFactory)
```

## Alias

You can optionally specify an alias for the ORM Entity:

```js
racer.orm('players.*', PlayerModel, 'Player')
```

This will allow you to explicitly specify in `.at()` and `.scope()`
which ORM Entity to use even for the unknown path patters:

```js
let playerId = 'playerId1'
// will create the PlayerModel, since it matches the specified path pattern:
model.scope('players.' + playerId).alert('please join the game!')

// The following will also create the PlayerModel
// even though '_session.myPlayer' wasn't specified in the orm path patterns:
model.scope('_session.myPlayer', 'Player').alert('please join the game!')
```

**IMPORTANT:** Note, that this is a bad practice and must only be used in the edge cases.

It's always better to list all your path patterns explicitly and don't use aliases at all:

```js
racer.orm('players.*', PlayerModel)
racer.orm('_session.myPlayer', PlayerModel)
racer.orm('_session.rivalPlayer', PlayerModel)
```

## Licence

MIT

(c) Decision Mapper - http://decisionmapper.com
