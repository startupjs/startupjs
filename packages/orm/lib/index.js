const Model = require('racer').Model
const promisifyRacer = require('./promisifyRacer')

module.exports = exports = function (racer) {
  if (racer.orm) return

  racer._orm = {}

  racer.orm = function (pattern, OrmEntity, alias) {
    var name = alias || pattern
    if (racer._orm[name]) throw alreadyDefinedError(pattern, alias)
    racer._orm[name] = {
      pattern: pattern,
      regexp: patternToRegExp(pattern),
      OrmEntity: OrmEntity
    }
  }

  Model.prototype.at = function (subpath, alias) {
    var path = this.path(subpath)
    return this.scope(path, alias)
  }

  Model.prototype._scope = function (path) {
    var ChildModel = Model.ChildModel
    var model = new ChildModel(this)
    model._at = path
    return model
  }

  Model.prototype.scope = function (path, alias) {
    if (alias) {
      if (racer._orm[alias]) {
        return this.__createScopedModel(path, racer._orm[alias].OrmEntity)
      } else {
        throw new Error(
          'Non-existent alias of the OrmEntity specified: ' + alias + '\n\n' +
          'Most likely you have specified the path incorrectly in ' +
          '".scope()" or ".at()"\n\n' +
          'The path must be passed as a single string separated by dots, ' +
          'for example:\n\n' +
          'CORRECT:\n' +
          '$root.at(\'users.\' + userId)\n\n' +
          'INCORRECT:\n' +
          '$root.at(\'users\', userId)'
        )
      }
    }

    var segments = this._dereference(this.__splitPath(path), true)
    var fullPath = segments.join('.')

    for (var name in racer._orm) {
      var regexp = racer._orm[name].regexp
      if (regexp.test(fullPath)) {
        return this.__createScopedModel(path, racer._orm[name].OrmEntity)
      }
    }

    return this._scope(path)
  }

  Model.prototype.__createScopedModel = function (path, OrmEntity) {
    var model
    if (OrmEntity.factory || OrmEntity.prototype.factory) {
      model = OrmEntity(this._scope(path), this)
      // if factory didn't return anything, return a simple scoped model
      if (!model) model = this._scope(path)
    } else {
      model = new OrmEntity(this)
    }
    model._at = path
    return model
  }

  Model.prototype.__splitPath = function (path) {
    return (path && path.split('.')) || []
  }

  promisifyRacer()
}

function patternToRegExp (pattern) {
  pattern = pattern.replace(/\$/g, '\\$').replace(/\./g, '\\.').replace(/\*/g, '([^\\.]*)')
  return new RegExp('^' + pattern + '$')
}

function alreadyDefinedError (pattern, alias) {
  var msg
  if (alias) {
    msg = 'ORM entity with the alias \'' + alias + '\' is already defined. ' +
      'Aliases must be unique. If you did already define the same ORM entity with ' +
      'that alias name, just don\'t specify the alias at all -- path pattern is sufficient.'
  } else {
    msg = 'ORM entity matching the same path pattern \'' + pattern + '\' is already defined.'
  }
  return new Error(msg)
}

exports.ChildModel = Model.ChildModel

function BaseModel () {
  Model.ChildModel.apply(this, arguments)
}
BaseModel.prototype = Object.create(Model.ChildModel.prototype)
BaseModel.prototype.constructor = BaseModel

BaseModel.prototype.getId = function () {
  var actualField = this.dereferenceSelf()
  return actualField.leaf()
}

BaseModel.prototype.getCollection = function () {
  var model = this.root
  var actualField = this.dereferenceSelf()
  return model._splitPath(actualField.path())[0]
}

BaseModel.prototype.dereferenceSelf = function () {
  var model = this.root
  var segments = model._splitPath(this.path())
  return model.scope(model._dereference(segments, true).join('.'))
}

exports.BaseModel = BaseModel
