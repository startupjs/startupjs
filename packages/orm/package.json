{
  "name": "@startupjs/orm",
  "type": "module",
  "version": "0.40.0",
  "description": "ORM system for Racer.js and ShareDB",
  "main": "lib/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "pluralize": "^8.0.0",
    "racer": "1.0.0"
  },
  "author": "cray0000 <cray0000@gmail.com>",
  "license": "MIT",
  "keywords": [
    "racer",
    "sharedb",
    "share",
    "plugin"
  ],
  "readmeFilename": "README.md"
}
