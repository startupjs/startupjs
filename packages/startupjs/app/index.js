module.exports = require('@startupjs/app')
