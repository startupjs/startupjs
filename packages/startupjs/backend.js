export { default } from '@startupjs/backend'
export * from '@startupjs/backend'
