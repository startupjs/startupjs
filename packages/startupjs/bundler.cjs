throw new Error(`
[startupjs/bundler] ERROR! StartupJS bundler api has changed.

Now instead of using general \`require('startupjs/bundler.cjs')\` you have to require
a specific file you need.

Here is what you need to change:

1. \`require('startupjs/bundler.cjs').webpackWebConfig\`
    v v v
   \`require('startupjs/bundler/webpack.web.config.cjs')\`

2. \`require('startupjs/bundler.cjs').webpackServerConfig\`
    v v v
   \`require('startupjs/bundler/webpack.server.config.cjs')\`

3. \`require('startupjs/bundler.cjs').babelConfig\`
    v v v
   \`require('startupjs/bundler/babel.config.cjs')\`

4. \`require('startupjs/bundler.cjs').metroConfig\`
    v v v
   \`require('startupjs/bundler/metro.config.cjs')\`

5. \`require('startupjs/bundler.cjs').rnConfig\`
    v v v
   \`require('startupjs/bundler/react-native.config.cjs')\`

6. \`require('startupjs/bundler.cjs').viteConfig\`
    v v v
   \`require('startupjs/bundler/vite.config.cjs')\`

Additionally, if you did override .js/.jsx compilation in your webpack.web.config.cjs, you
might have used \`getJsxRule\` function there. In which case you should now get it from helpers:

7. \`const { getJsxRule } = require('startupjs/bundler.cjs')\`
    v v v
   \`const { getJsxRule } = require('startupjs/bundler/helpers.cjs')\`
`)
