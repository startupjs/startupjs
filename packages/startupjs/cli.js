#!/usr/bin/env node
'use strict'

import cli from '@startupjs/cli'
import { fileURLToPath } from 'url'
import path, { dirname } from 'path'

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

const templatesPath = path.join(__dirname, 'templates')

cli.run({ templatesPath })
