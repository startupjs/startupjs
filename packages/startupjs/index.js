export * from '@startupjs/react-sharedb'
export * from '@startupjs/hooks'
// HINT: `isomorphic` means that the code can be executed both
//        on the server and on the client
export * from '@startupjs/isomorphic-helpers'
// dummy babel macro functions for @startupjs/babel-plugin-rn-stylename-inline
export function css () {}
export function styl () {}
export { default as t } from '@startupjs/i18n/client/t.js'
