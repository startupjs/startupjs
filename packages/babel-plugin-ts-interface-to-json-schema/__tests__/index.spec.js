const pluginTester = require('babel-plugin-tester').default
const plugin = require('../index')
const { name: pluginName } = require('../package.json')

pluginTester({
  plugin,
  pluginName,
  snapshot: true,
  pluginOptions: {},
  tests: [
    {
      title: 'It is convert typescript interface to JSON-Schema and add that JSON to ORM model as static field',
      code: `
        import { BaseModel } from 'startupjs/orm'
    
        interface IEventModel {
          name: string;
          /**
          * @default [1.23, 65.21, -123.40, 0, 1000000.0001]
          */
          amount: number[];
        }
        
        export default class EventModel extends BaseModel {
          methodOne() {
            return 1
          }
          methodTwo() {
            return 2
          }
        }`,
      snapshot: true
    }
  ]
})
