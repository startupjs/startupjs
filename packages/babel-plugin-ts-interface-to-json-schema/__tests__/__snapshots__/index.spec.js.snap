// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@startupjs/babel-plugin-ts-interface-to-json-schema It is convert typescript interface to JSON-Schema and add that JSON to ORM model as static field: It is convert typescript interface to JSON-Schema and add that JSON to ORM model as static field 1`] = `

import { BaseModel } from 'startupjs/orm'
    
interface IEventModel {
  name: string;
  /**
  * @default [1.23, 65.21, -123.40, 0, 1000000.0001]
  */
  amount: number[];
}

export default class EventModel extends BaseModel {
  methodOne() {
    return 1
  }
  methodTwo() {
    return 2
  }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { BaseModel } from 'startupjs/orm'
interface IEventModel {
  name: string;
  /**
   * @default [1.23, 65.21, -123.40, 0, 1000000.0001]
   */

  amount: number[];
}
export default class EventModel extends BaseModel {
  static schema = {
    type: 'object',
    properties: {
      name: {
        type: 'string'
      },
      amount: {
        default: [1.23, 65.21, -123.4, 0, 1000000.0001],
        type: 'array',
        items: {
          type: 'number'
        }
      }
    },
    $schema: 'http://json-schema.org/draft-07/schema#'
  }

  methodOne() {
    return 1
  }

  methodTwo() {
    return 2
  }
}


`;
